### PR TITLE
[ENH] basic setar-tree module and tests

### PR DESCRIPTION
**Pull request by TinaJin0228**
#### Reference Issues/PRs
#2816. 

#### What does this implement/fix? Explain your changes.

In this commit, the major functions for the setar-tree algorithm is implemented, and the basic test file to test them. 

#### Does your contribution introduce a new dependency? If yes, which one?

No. 

#### Any other comments?

This is the initial commit for #2816.

Next steps:
- Elaborate on the SETAR-tree algorithm, including an accelerated grid-search for finding the optimal split.  
- Add tests for corner cases and real-world datasets.  
- Write documentation and update the API reference page.  
- Explore implementing separate TAR and SETAR forecasters.

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

##### For new estimators and functions
- [ ] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

##### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

### CODE REVIEW

File: aeon/forecasting/_setar.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    This implementation is based on the logic from the `get_setar_forecasts`
+    function in the original paper's R code
+    (https://github.com/rakshitha123/SETAR_Trees).
+
+    Parameters
+    ----------
+    lag : int, default=10
+        The maximum number of past lags to consider for both the AR models
+        and as the thresholding variable.
```
Comment (MatthewMiddlehurst) [inline aeon/forecasting/_setar.py]: You should document the horizon here also, you can reuse the text from other classes..

File: aeon/forecasting/_setar.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    """
```
Comment (MatthewMiddlehurst) [inline aeon/forecasting/_setar.py]: I would add an example usage of the class to the docstring here.

File: aeon/forecasting/_setar.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        return np.fliplr(np.array(X_list)), np.array(y_list)
+
+    def _fit(self, y, exog=None):
+        """Fit the SETAR model to a single time series."""
+        self._last_window = y[-self.lag :].copy()
+        best_overall_sse = float("inf")
+        best_model_params = None
+
+        for _lag in range(self.lag, 0, -1):
+            if len(y) <= _lag:
```
Comment (MatthewMiddlehurst) [inline aeon/forecasting/_setar.py]: just `lag` should be fine for this loop, no _ needed as its not an attribute.

File: aeon/forecasting/_setar.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
+"""SETAR: A classic univariate forecasting algorithm."""
+
+import numpy as np
+from sklearn.linear_model import LinearRegression
+
+from aeon.forecasting.base import BaseForecaster
+
+
+class SetarForecaster(BaseForecaster):
+    """
```
Comment (MatthewMiddlehurst) [inline aeon/forecasting/_setar.py]: Could you please add a test file for this class and add some functions. If you could generate some expected results from another implementation to ensure correctness that would be good as well.

Comment (MatthewMiddlehurst) [inline aeon/forecasting/_setar.py]: IMO the class name should just be `SETAR`. I do not think this is used anywhere other than forecasting, and we can easily change if it is.

File: aeon/forecasting/__init__.py
Comment (MatthewMiddlehurst) [inline aeon/forecasting/__init__.py:1]: Add the regular SETAR to the init as well.

**Review by MatthewMiddlehurst — CHANGES_REQUESTED** [review]
Some comments, more to come.

File: aeon/forecasting/_setartree.py [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+from sklearn.linear_model import LinearRegression
+
+from aeon.forecasting.base import BaseForecaster
+
+__maintainer__ = ["TinaJin0228"]
+__all__ = ["SetartreeForecaster"]
+
+
+class SetartreeForecaster(BaseForecaster):
+    """
```
Comment (MatthewMiddlehurst) [inline aeon/forecasting/_setartree.py]: I think `SETARTree` is Bette as a name. Same reason as SETAR.

**Review by MatthewMiddlehurst — COMMENTED** [review]
I think this is a good start. Could you post here on what your current plans are for testing correctness i.e. data used and results/implementation to compare against.

You mentioned previously having some issues with implementing global methods using the current framework. Could you also post that here?

Please add the classes to the API documentation in `docs/`

File: aeon/forecasting/_setartree.py (hunk 5 above)
Comment (TinaJin0228) [reply aeon/forecasting/_setartree.py]: I will also modify SetarforestForecaster to SETARForest

**Comment by TinaJin0228** [comment]
Current plan for testing: 
Several experiment results are in the blog (https://medium.com/@jintina48/gsoc-experiment-record-f0aa3bd82c18). 
The function of fitting and forecasting is ok, but there is still a significant gap of the results, so my current plan is:
1. Carefully compare my implementation to that of the R codebase, especially the splitting function and the error calculator (which, in my opinion, is the most likely one to have flaws)(though I need more time to locate them)
2. Insist on first testing on the Chaotic dataset, since it is the simplest and uniform dataset in the paper's evaluation benchmark. 

Issues implementing global methods using the current framework: 
The current Aeon framework treats multiple time series input as multivariate time series. 
Current temporary solution is to pass one time series to the input "x" and others to the exogenous variable, which is a quick fix. 

As for the efficiency: 
I think the primary computational bottleneck is the `find_optimal_split` function, which is the major step in building the tree.

File: aeon/forecasting/_setartree.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    significance_divider : int, default=2
+        The factor by which to divide the significance level at each tree depth.
+    error_threshold : float, default=0.03
+        The minimum percentage of error reduction required to make a split.
+    """
+
+    def __init__(
+        self,
+        lag: int = 10,
+        horizon: int = 1,
```
Comment (TonyBagnall) [inline aeon/forecasting/_setartree.py]: for now, set horizon tag to false and dont p[ass the horizon

File: aeon/forecasting/_setartree.py [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+
+        # Fit models for all leaf nodes
+        for node_id, node in self.tree_.items():
+            if node["is_leaf"] or node.get("left_child") is None:
+                self.leaf_models_[node_id] = self._fit_pr_model(node["X"], node["y"])
+                node["is_leaf"] = True
+
+        return self
+
+    def _predict(self, y=None, exog=None):
```
Comment (TonyBagnall) [inline aeon/forecasting/_setartree.py]: dont need to check is fitted here, its done in the base class

File: aeon/forecasting/_setartree.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        else:
+            # Ensure y is 2D
+            if y.ndim == 1:
+                y = y.reshape(1, -1)
+            # We take the last 'lag' points of y for the initial prediction
+            history = y[0, -self.lag :]
+
+        predictions = []
+
+        for _ in range(self.horizon):
```
Comment (TonyBagnall) [inline aeon/forecasting/_setartree.py]: dont heed a horizon here. predict should simply predict one ahead based on y. Further prediction horizon predictions are made with the iterative_forecast()

**Review by TonyBagnall — CHANGES_REQUESTED** [review]

**Comment by TinaJin0228** [comment]
Progress: I’ve successfully reproduced the results from the paper (with a tiny gap) on the Chaotic dataset, using an independent Python implementation. I’ll update the Aeon branch once I’ve confirmed that the method works within the Aeon framework. Together be done alongside the modification requests mentioned above.

File: aeon/forecasting/_setarforest.py [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+
+__maintainer__ = ["TinaJin0228"]
+__all__ = ["SETARForest"]
+
+
+class SETARForest(BaseForecaster):
+    """
+    SETAR-Forest: An ensemble of SETAR-Trees for global time series forecasting.
+
+    This implementation is based on the paper "SETAR-Tree: a novel and accurate
```
Comment (TonyBagnall) [inline aeon/forecasting/_setarforest.py]: add a references section and put the reference there

**Comment by TonyBagnall** [comment]
@TinaJin0228 I cant check directly in code, but these tests and the doctest are failing

You need to set capability:multivariate to False gets rid of most of them, the
"IndexError: index 0 is out of bounds for axis 0 with size 0" not sure about

doctest is failing for the example, probably to do with the output type, I recommend just removing the example for now if you cant easily fix it

FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_fit_deterministic(estimator=SETARTree(),datatype=UnivariateSeries-None)] - ValueError: Multivariate data not supported by SETARTree
FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_forecaster_output(estimator=SETARForest(),datatype=UnivariateSeries-None)] - IndexError: index 0 is out of bounds for axis 0 with size 0
FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_fit_updates_state_and_cloning(estimator=SETARForest(),datatype=UnivariateSeries-None)] - IndexError: index 0 is out of bounds for axis 0 with size 0
FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_forecaster_output(estimator=SETARTree(),datatype=UnivariateSeries-None)] - ValueError: Multivariate data not supported by SETARTree
FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_non_state_changing_method(estimator=SETARTree(),datatype=UnivariateSeries-None)] - ValueError: Multivariate data not supported by SETARTree
FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_persistence_via_pickle(estimator=SETARForest(),datatype=UnivariateSeries-None)] - IndexError: index 0 is out of bounds for axis 0 with size 0
FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_fit_deterministic(estimator=SETARForest(),datatype=UnivariateSeries-None)] - IndexError: index 0 is out of bounds for axis 0 with size 0
FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_persistence_via_pickle(estimator=SETARTree(),datatype=UnivariateSeries-None)] - ValueError: Multivariate data not supported by SETARTree
FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_fit_updates_state_and_cloning(estimator=SETARTree(),datatype=UnivariateSeries-None)] - ValueError: Multivariate data not supported by SETARTree
FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_non_state_changing_method(estimator=SETARForest(),datatype=UnivariateSeries-None)] - IndexError: index 0 is out of bounds for axis 0 with size 0

**Comment by TinaJin0228** [comment]
https://docs.google.com/spreadsheets/d/1e0nw01LSm4cGylJxWgKaozj3O50eij5CqU-buSG_TtU/edit?gid=0#gid=0
This sheet records experiments on three datasets (chaotic, Mackey-Glass, and M5), comparing two versions of code (the author’s R codebase and a Python version under the aeon framework) against the claimed results in the paper. The tests use three tree-building stopping criteria: lin_test, error_imp, or both. The first part of the results is from the paper, the second part is reproduced from the author’s codebase, and the third is from the Python module in aeon. The results for the M5 dataset in the reproduced R section are omitted since the code failed.

The results show that:

Comparing our aeon implementation to the paper’s claim, its performance depends heavily on the stopping criteria. When using the both criteria, the aeon implementation is competitive and works well. However, when using either lin_test or error_imp alone, the aeon implementation is generally worse than the paper's results, especially on the Chaotic dataset.

Surprisingly, the result for the msMAPE metric on the M5 dataset was much better than the paper's claim, but I haven't figured out why yet.

Most results in the paper can be reproduced using the author's R code. The findings are largely consistent, but with a notable exception: the reproduced mean MASE for the error_imp criteria was significantly higher than what the paper reported.

**Comment by TonyBagnall** [comment]
its quite hard to tell why this is failing because I cant clone the repo, but I suspect exogenous checks are triggering this? 

FAILED aeon/testing/tests/test_all_estimators.py::test_all_estimators[check_fit_deterministic(estimator=SETARTree(),datatype=UnivariateSeries-None)] - ValueError: Multivariate data not supported by SETARTree

**Comment by TinaJin0228** [comment]
new refinements to setar & setar-forest today. Unit tests to setar module are added. 
Regarding the multivariate error (ValueError: Multivariate data not supported by SETARTree/SETARForest), I’m unsure of the cause. Aeon’s documentation states that “Passing a multivariate array will raise an error if capability:multivariate is False,” but I haven’t been able to identify where I might be mistakenly passing a multivariate array to these modules.

ps: I still strongly recommend that Aeon’s forecasting module support global time series forecasting. The current BaseForecaster is designed for fitting on a single time series, and using exog to pass additional target series feels like a temporary workaround rather than a robust solution.

**Comment by TonyBagnall** [comment]
> new refinements to setar & setar-forest today. Unit tests to setar module are added. Regarding the multivariate error (ValueError: Multivariate data not supported by SETARTree/SETARForest), I’m unsure of the cause. Aeon’s documentation states that “Passing a multivariate array will raise an error if capability:multivariate is False,” but I haven’t been able to identify where I might be mistakenly passing a multivariate array to these modules.
> 
> ps: I still strongly recommend that Aeon’s forecasting module support global time series forecasting. The current BaseForecaster is designed for fitting on a single time series, and using exog to pass additional target series feels like a temporary workaround rather than a robust solution.

so afaik global forecasting has two meanings

1. forecast a single series using multiple series. This is the univariate y/exogenous model. Regression supports this
2. forecast multiple series using multiple series. This is basically multivariate forecasting, capability:multivariate. This is pass a single y shape (n_channels, n_timepoints) and predict (n_channels) for a given horizon. Currently we dont have any algorithms that do that, and would need to slightly rework the interface. We could actually implement a broadcaster to do this with arma models etc independently on each channel, thats a solid benchmark

File: aeon/forecasting/_setar.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        best_sse = np.inf
+        best_th = None
+        best_inter_low = None
+        best_coefs_low = None
+        best_inter_high = None
+        best_coefs_high = None
+
+        # X for AR: lags 1..l (plus intercept)
+        # X = add_constant(lagged[:, :l])
+        X = add_constant(lagged[:, :lag_order], has_constant="add")
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: you can do this with np.hstack and remove the need for statsmodels

File: aeon/forecasting/_setar.py [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        # set one-step-ahead forecast_ for forecast()
+        self.forecast_ = float(self._one_step_from_tail(y))
+        return self
+
+    def _fit_setar(self, y, lag_order, delay):
+        maxlag = lag_order + delay
+        if len(y) <= maxlag:
+            raise ValueError(f"Series too short for lag {lag_order}.")
+
+        lagged = lagmat(y, maxlag, trim="both")  # shape: (T-maxlag, maxlag)
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: again, you can do this in numpy easily enough without statsmodels, chatgpt gives me this, obvs needs debugging

def lagmat_numpy(x, maxlag):
    x = np.asarray(x)
    if x.ndim == 1:
        x = x[:, None]  # make it 2D
    n_obs = x.shape[0]
    return np.hstack([x[maxlag - i:n_obs - i] for i in range(maxlag + 1)])

File: aeon/forecasting/_setar.py [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+                pass
+
+        if not fitted:
+            # Fallback to linear regression AR(self.lag)
+            maxlag = self.lag
+            if len(y) <= maxlag:
+                raise ValueError("Series too short for fallback fitting.")
+            lagged = lagmat(y, maxlag, trim="both")  # cols: y_{t-1}..y_{t-maxlag}
+            # X = add_constant(lagged)
+            X = add_constant(lagged, has_constant="add")
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: you can do this without statsmodels with hstack

File: aeon/forecasting/_setar.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+            except Exception:
+                # try a smaller lag
+                pass
+
+        if not fitted:
+            # Fallback to linear regression AR(self.lag)
+            maxlag = self.lag
+            if len(y) <= maxlag:
+                raise ValueError("Series too short for fallback fitting.")
+            lagged = lagmat(y, maxlag, trim="both")  # cols: y_{t-1}..y_{t-maxlag}
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: or this 


def lagmat_strided(x, maxlag):
    x = np.asarray(x)
    if x.ndim == 1:
        x = x[:, None]
    n_obs, n_vars = x.shape
    shape = (n_obs - maxlag, maxlag + 1, n_vars)
    strides = (x.strides[0], -x.strides[0], x.strides[1])
    rolled = as_strided(x[maxlag:], shape=shape, strides=strides)
    return rolled.reshape(n_obs - maxlag, -1)

File: aeon/forecasting/_setar.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
+"""SETAR forecaster with 2 regimes and fallback to linear regression."""
+
+__maintainer__ = ["TinaJin0228"]
+__all__ = ["SETAR"]
+
+import numpy as np
+from statsmodels.regression.linear_model import OLS
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: if you want to use statsmodels you need to add it as a soft dependency and do the import in the function. It would be better to use an alternative to OLS with numpy/scipy if possible

File: aeon/forecasting/_setar.py [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+                continue
+
+            X_low = X[low_idx]
+            y_low = trimmed_y[low_idx]
+            X_high = X[high_idx]
+            y_high = trimmed_y[high_idx]
+
+            model_low = OLS(y_low, X_low).fit()
+            model_high = OLS(y_high, X_high).fit()
+            sse = model_low.ssr + model_high.ssr
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: if all you need is the SSR rather than store the model it can be done more efficiently in numpy or bespoke numba

File: aeon/forecasting/_setarforest.py [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+            n_thresholds=self.n_thresholds,
+            scale=self.scale,
+            seq_significance=self.seq_significance,
+        )
+        training_set = template._process_input_data(y, exog)
+        embedded_df, _, _ = template._create_tree_input_matrix(training_set)
+        return embedded_df
+
+    def _fit(self, y, exog=None):
+        # Full embedded matrix (reuses SETARTree logic)
```
Comment (TonyBagnall) [inline aeon/forecasting/_setarforest.py]: is there a reason you are putting it all in a dataframe? Its less efficient

Comment (TinaJin0228) [reply aeon/forecasting/_setarforest.py]: not really, I just did not realize that data frame is slow...

**Comment by TinaJin0228** [comment]
> personally I'm not a fan of embedding dataframes inside an algorithm, it slows it down massively and I'm not sure why its required, but thats just my opinion

I have created a statsmodels-free SETAR module and a matching test file.

**Comment by TonyBagnall** [comment]
looks like its just this odd multivariate fail, I think that might be a base class issue, I'll take a look

**Comment by TonyBagnall** [comment]
I think its failing because setar forest does not call the base class constructor

        super().__init__(horizon=1, axis=0)

File: aeon/forecasting/_setar.py [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+    lag : int, default=10
+        The maximum autoregressive order (embedding dimension) to attempt.
+    """
+
+    _tags = {
+        "capability:horizon": False,
+        "y_inner_type": "np.ndarray",
+    }
+
+    def __init__(self, lag=10, horizon=1):
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: I dont think you actually use the horizon in the algorithm, so dont have it as a parameter, I dont think it does anything

File: aeon/forecasting/_setar.py [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+            self.fallback_coefs = coefs
+            self.model = "linear"
+            self.current_lag = len(self.fallback_coefs)
+
+        # set one-step-ahead forecast_ for forecast()
+        self.forecast_ = float(self._one_step_from_tail(y))
+        return self
+
+    def _fit_setar(self, y, lag_order, delay):
+        maxlag = lag_order + delay
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: test this once in fit, if its ok on the first iteration its ok for them all

File: aeon/forecasting/_setar.py [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+            if sse < best_sse:
+                best_sse = sse
+                best_th = th
+                best_inter_low = inter_l
+                best_coefs_low = coef_l
+                best_inter_high = inter_h
+                best_coefs_high = coef_h
+
+        if best_th is None:
+            raise ValueError("Could not find a suitable threshold.")
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: given this is caught above, better to just return None and continue if None returned. Exceptions make it harder to JIT

File: aeon/forecasting/_setar.py [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+
+        return best_th, best_coefs_low, best_inter_low, best_coefs_high, best_inter_high
+
+    def _one_step_from_tail(self, y_1d):
+        if self.current_lag is None:
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: dont test these here, need to optimise ths function

File: aeon/forecasting/_setar.py [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+            self.fallback_intercept = inter
+            self.fallback_coefs = coefs
+            self.model = "linear"
+            self.current_lag = len(self.fallback_coefs)
+
+        # set one-step-ahead forecast_ for forecast()
+        self.forecast_ = float(self._one_step_from_tail(y))
+        return self
+
+    def _fit_setar(self, y, lag_order, delay):
```
Comment (TonyBagnall) [inline aeon/forecasting/_setar.py]: this can be a static method I think?

File: aeon/forecasting/machine_learning/_setartree.py [hunk 22]
```diff
@@ -0,0 +1,508 @@
...
+        self.significance = significance
+        self.significance_divider = significance_divider
+        self.error_threshold = error_threshold
+        self.n_thresholds = n_thresholds
+        self.scale = scale
+        self.seq_significance = seq_significance
+        self.fixed_lag = fixed_lag
+        self.external_lag = external_lag
+        self.feature_subset = feature_subset
+        super().__init__(horizon=1, axis=1)
```
Comment (TonyBagnall) [inline aeon/forecasting/machine_learning/_setartree.py]: I'd call this at the end of the constructor myself

**Review by TonyBagnall — CHANGES_REQUESTED** [review]
